scvi/autotune/_autotune.py

```diff
@@ -0,0 +1,307 @@
+        trainer_hyperparams: Optional[dict] = None,
+        plan_hyperparams: Optional[dict] = None,
+        setup_args: Optional[dict] = None,
+        continuous_covariates: Optional[List[str]] = None,
+        categorical_covariates: Optional[List[str]] = None,
+        test_effect_covariates: bool = False,
+        test_effect_hvg: bool = False,
+        top_hvg: Optional[List[int]] = None,
+        batch_key_hvg: Optional[str] = None,
+    ):
+        training_metrics = training_metrics or []
+        metric_functions = metric_functions or {}
+        model_hyperparams = model_hyperparams or {}
+        trainer_hyperparams = trainer_hyperparams or {}
+        plan_hyperparams = plan_hyperparams or {}
+        setup_args = setup_args or {}
+        continuous_covariates = continuous_covariates or []
+        categorical_covariates = categorical_covariates or []
+        top_hvg = top_hvg or [None]
+
+        if test_effect_hvg and not top_hvg:
+            raise ValueError(
+                "test_effect_hvg is set to True but no list of top_hvgs was given. Please"
+                " provide a list"
+            )
+
+        self.adata = adata
+        self.model = model
+        self.setup_signature = inspect.signature(
+            self.model.setup_anndata
+        ).parameters  # Fetch available parameters
+        self.training_metrics = training_metrics
+        self.metric_functions = metric_functions
+        self.model_hyperparams = model_hyperparams
+        self.trainer_hyperparams = trainer_hyperparams
+        self.plan_hyperparams = plan_hyperparams
+        self.setup_args = setup_args
+
+        continuous_covariates = [[covariate] for covariate in continuous_covariates]
+        categorical_covariates = [[covariate] for covariate in categorical_covariates]
+        if test_effect_covariates:
+            continuous_covariates.append(None)
+            categorical_covariates.append(None)
+
+        self.continuous_covariates = {
+            "continuous_covariate_keys": tune.choice(continuous_covariates)
+            if continuous_covariates
+            and "continuous_covariate_keys"
+            in self.setup_signature  # Change default None only if setup_anndata signature has covariates
+            else None
+        }
+        self.categorical_covariates = {
+            "categorical_covariate_keys": tune.choice(categorical_covariates)
+            if categorical_covariates
+            and "categorical_covariate_keys"
+            in self.setup_signature  # Change default None only if setup_anndata signature has covariates
+            else None
+        }
+
+        if test_effect_hvg:
+            self.test_effect_hvg = {"subset_adata": tune.choice([True, False])}
+        else:
+            self.test_effect_hvg = {"subset_adata": False}
+
+        # Using conditional search spaces is not supported in all space search algorithms
+        # If necessary we would need to re-factor to Optuna / HyperOptSearch spaces
```

Comment:
Yes, of course. It's a bit complicated to grasp but I hope I can explain it well.

First, `test_effect_hvg` can be set by the user to either `True` or `False`. If it is set to `True` we want to reflect this in the search space with the `self.test_effect_hvg` dictionary key. Since we want to test the effect of filtering for HVGs, we randomly generate `True` or `False` values that will be assigned to `self.test_effect_hvg`. If the user sets the `test_effect_hvg` to `False` ,  `self.test_effect_hvg` will be just set to `False` without using ray search spaces. 

Here the conditional bit is found in lines 142-148:

```
 self.top_hvg = {
            "top_n": tune.sample_from(
                lambda spec: np.random.choice(top_hvg)
                if spec.config["subset_adata"]
                else None
            )
        }
```
See: [ray tune conditional search spaces](https://docs.ray.io/en/master/tune/tutorials/tune-search-spaces.html#how-to-use-custom-and-conditional-search-spaces). Here `if spec.config["subset_adata"]` is checking the value of this variable in the search space **while** it is being created. 

I think it's easier to understand if we imagine the case without conditional search spaces. In this case the snippet would look like this:

```
 self.top_hvg = {
            "top_n": tune.sample_from(
                lambda spec: np.random.choice(top_hvg)
                if self.test_effect_hvg
                else None
            )
        }
```

Note the use of `self.test_effect_hvg` instead of `spec.config["subset_adata"]`. In this case, if `self.test_effect_hvg` is set to `True` (i.e. the user wants to test the effect of hvg filtering) then self.top_hvg will always be set to a given number of `top_hvg`, even though we would like to test also the effect of not filtering. When using `spec.config["subset_adata"]` is checking the value of this variable in the search space **while** it is being created, so even if the user sets `test_hvg_effect` to `True`, the search space will still consider not filtering for `top_hvg`. 

I reckon it is a bit too much over text, let me know if we need to discuss this further

Reply:
thanks for the clarification, we can stick to using this implementation for now and refactor if necessary